include: Prefix includes to use a scope

yperess opened the pull request:
Move include paths and add new target_include_directories to support
backwards compatibility:
* /include -> /include/zephyr
  example: <irq.h> -> <zephyr/irq.h>
* /kernel/include -> /kernel/include/zephyr/kernel/
  example: <thread.h> -> <zephyr/kernel/threads.h>

Issue #41543

Signed-off-by: Yuval Peress <peress@google.com>

Commit:
include: Prefix includes to use a scope

Move include paths and add new target_include_directories to support
backwards compatibility:
* /include -> /include/zephyr
  example: <irq.h> -> <zephyr/irq.h>

Issue #41543

Signed-off-by: Yuval Peress <peress@google.com>

andyross reviewed (dismissed):
Ooooph....  standard Grumpy Curmudgeon complaints apply here: touching every file in the project makes patch rebasing across versions a pain, affects users with portability concerns, and generally just leads to headache.

That said, yes, our include structure is a mess.   Yes, it's probably overdue for a cleanup.  And yes, I like this scheme better than what we have (rather, "the most charitable interpretation of the design intent of what we have", heh).  So I guess this is +1 from me.

But can we please make sure this is what we really want and that we're sure we're not just going to need to do this again next bikeshed season?  I still bear scars from our "int -> u32_t -> uint32_t" epoch.

yperess commented:
> Ooooph.... standard Grumpy Curmudgeon complaints apply here: touching every file in the project makes patch rebasing across versions a pain, affects users with portability concerns, and generally just leads to headache.
> 
> That said, yes, our include structure is a mess. Yes, it's probably overdue for a cleanup. And yes, I like this scheme better than what we have (rather, "the most charitable interpretation of the design intent of what we have", heh). So I guess this is +1 from me.
> 
> But can we please make sure this is what we really want and that we're sure we're not just going to need to do this again next bikeshed season? I still bear scars from our "int -> u32_t -> uint32_t" epoch.

100% with you, I think that once I get this working fully and passing the tests it'll be easy to tweak. So if you have any preferenmces please let me know.

yperess commented:
@mbolivar-nordic can we remove the DNM label now that the TSC voted?

gmarull commented:
+1

fabiobaltieri commented on CMakeLists.txt:
```diff
@@ -106,6 +106,7 @@ zephyr_library_named(zephyr)
 
 zephyr_include_directories(
   include
+  include/zephyr
```
Should this be conditional to a Kconfig option? That way we can leave a compatibility mode, but also have an option to help migrate downstream projects or ensure that we don't reintroduce old path includes after this.

gmarull replied:
I agree we should have a Kconfig option. Other things I'd consider, given that a future removal of `include/` option will break the compilation of every single file:

- Deprecate current `include/` path, but keep it the default except for upstream code
- Add a big warning when building with the "compatible" mode until it is not deleted (users will notice this way)
- Maybe provide a migration script that automatically prefixes every Zephyr header used in apps/libs with `zephyr/`

yperess replied:
Yes, I just wanted to get the CI to pass first before I add another layer to it.

Commit:
build: make non zephyr/ include prefix configurable

Add a Kconfig to disable non prefixed includes. By setting
`CONFIG_LEGACY_INCLUDE_PATH=n` developers can disable having
`include/zephyr` in the search path. This means that includes such
as `#include <kernel.h>` will no longer work.

Internally, every header should be updated to add the `zephyr/`
prefix to the headers. Only then, will developers be able to use
this config value for their applications.

Signed-off-by: Yuval Peress <peress@google.com>

Commit:
xtensa: Add types to kconfigs

Fix a documentation build warning by adding the type to the configs

Signed-off-by: Yuval Peress <peress@google.com>

yperess commented:
I'm not quite sure why the build shows up as broken, locally I'm able to pass: `$ ./scripts/twister -v -i -p nucleo_f429zi -s samples/posix/gettimeofday/sample.posix.gettimeofday`

stephanosio commented:
> I'm not quite sure why the build shows up as broken, locally I'm able to pass: `$ ./scripts/twister -v -i -p nucleo_f429zi -s samples/posix/gettimeofday/sample.posix.gettimeofday`

@yperess Have you tried `west update`? I am able to reproduce this issue on the latest main.

FYI @erwango @FRASTM

yperess commented:
OK, so currently everything passes except for the issues in https://github.com/zephyrproject-rtos/zephyr/issues/44486. I'll continue to rebase 2x a day to keep this PR relevant.

yperess commented:
@mbolivar-nordic can we remove the DNM label on this PR since the TSC voted that we can move forward with review for this change?

andyross reviewed (dismissed):
Refresh +1.  Obviously this is way too large to actually review in an exhaustive way manually, but the chosen include scheme is pleasing to my eyes and I can't find any mistakes in the execution.

yperess commented:
Rebased (no changes) @povergoing @cvinayak @andyross GitHub automatically dismisses the reviews, could you PTAL? Thanks, and I'm sorry for the inconvinience, hopefully we can merge this quickly.

henrikbrixandersen commented:
@erwango The failing CI may be of interest to you. It looks like it is related to the STM32 HAL.

erwango commented:
> @erwango The failing CI may be of interest to you. It looks like it is related to the STM32 HAL.

Fix for this issue just got merged. Can you rebase on latest main ?

yperess commented:
Rebased, @nordic-krch @povergoing @jhedberg I'm sorry, I need another review (hopefully this can merge before another conflict)

carlescufi commented:
@yperess one more time

nashif commented:
> /kernel/include -> /kernel/include/zephyr/kernel/

why is this needed given that kerne/include is internal?

andyross reviewed (dismissed):
Refresh +1

nashif reviewed (changes requested):
/kernel/include -> /kernel/include/zephyr/kernel/

why is this needed given that kerne/include is internal?

yperess commented:
> /kernel/include -> /kernel/include/zephyr/kernel/
> 
> why is this needed given that kerne/include is internal?

My thinking was to de-risk. While others may not `#include <zephyr/kernel/mmu.h>` there's nothing stopping an application from using another third party library that also has `mmu.h`. In that case (depending on how imports were done) even the internal build of Zephyr might confuse the Zephyr version of `mmu.h` and the other third party version. Also, it adds consistency.

stephanosio commented:
>  there's nothing stopping an application from using another third party library that also has mmu.h.

so there are two cases:

1. application code including the header files under `kernel/include`
2. kernel code including header files from the application(s)

The 1st case is not possible since `kernel/include` is only visible to the source files under `kernel/` and `arch/`.
The 2nd case should not be possible either since applications should limit the scope of their include exports to the application target(s).

In that sense, I agree that the `kernel/include` change should not be done.

nashif commented:
> My thinking was to de-risk. While others may not `#include <zephyr/kernel/mmu.h>` there's nothing stopping an application from using another third party library that also has `mmu.h`. In that case (depending on how imports were done) even the internal build of Zephyr might confuse the Zephyr version of `mmu.h` and the other third party version. Also, it adds consistency.

being able to  include <zephyr/kernel/mmu.h>  and being on the level and under the same namespace makes it look like this header is public and can be used by anyone.

yperess commented:
Looks like I missed a file at `include/zephyr/types.h` pushed again with that move

povergoing reviewed (dismissed):
+1 refresh
